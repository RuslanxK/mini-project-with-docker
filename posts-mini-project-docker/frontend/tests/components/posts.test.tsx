import { render, screen } from '@testing-library/react';
import Posts from '../../src/components/posts';
import { describe, it, expect } from 'vitest';
import { Post } from '../../src/interfaces/post';

// Mock data for the tests
const mockPost: Post = {
  _id: '1',
  title: 'First Post',
  description: 'This is the first post description.',
  createdAt: '2024-10-28T12:34:56Z',
};

describe('Posts Component', () => {
  // Test: Verifies that the loading message is displayed when the component is in a loading state
  it('should render the loading state', () => {
    render(<Posts postsData={[]} isLoading={true} />);
    const loadingElement = screen.getByText(/loading.../i);
    expect(loadingElement).toBeInTheDocument();
  });

  // Test: Verifies that the component displays a message when there are no posts
  it('should display a message when there are no posts', () => {
    render(<Posts postsData={[]} />);
    const noPostsMessage = screen.getByText(/no posts yet/i);
    expect(noPostsMessage).toBeInTheDocument();
  });

  // Test: Verifies that an error message is displayed when the component receives an error prop
  it('should display an error message when there is an error', () => {
    const error = new Error('Failed to fetch posts');
    render(<Posts postsData={[]} error={error} />);
    const errorElement = screen.getByText(/failed to fetch posts/i);
    expect(errorElement).toBeInTheDocument();
  });

  // Test: Checks if the component renders posts inside a list element
  it('should render posts inside a list structure', () => {
    render(<Posts postsData={[mockPost]} />);
    const list = screen.getByRole('list');
    expect(list).toBeInTheDocument();
  });


  // Test: Verifies that the post title and description are rendered correctly
  it('should render the post title and description correctly', () => {
    render(<Posts postsData={[mockPost]} />);
    
    const postTitle = screen.getByRole('heading', { name: /first post/i });
    expect(postTitle).toBeInTheDocument();

    const postDescription = screen.getByText(/this is the first post description/i);
    expect(postDescription).toBeInTheDocument();
  });

  // Test: Checks if the "Read More" button is rendered for each post
  it('should render the "Read More" button', () => {
    render(<Posts postsData={[mockPost]} />);
    const readMoreButton = screen.getByRole('button', { name: /read more/i });
    expect(readMoreButton).toBeInTheDocument();
  });

  // Test: Verifies that the post date is formatted and displayed correctly
  it('should format the date correctly', () => {
    render(<Posts postsData={[mockPost]} />);
    const formattedDate = screen.getByText('October 28, 2024');
    expect(formattedDate).toBeInTheDocument();
  });

  // Test: Verifies that the "Read More" button triggers the function when clicked
  it('should call the function when the "Read More" button is clicked', () => {
    const mockFunction = vi.fn();
    render(<Posts postsData={[mockPost]} onReadMore={mockFunction} />);
    
    const button = screen.getByRole('button', { name: /read more/i });
    button.click();
  
    expect(mockFunction).toHaveBeenCalledTimes(1);
  });
});
