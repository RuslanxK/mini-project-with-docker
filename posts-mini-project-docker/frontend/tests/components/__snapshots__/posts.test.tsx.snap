// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Posts Component > should match the snapshot 1`] = `
<div>
  <div
    class="max-w-full mx-auto px-4 sm:px-6 lg:px-8 py-16 bg-gray-100"
  >
    <h1
      class="text-5xl font-bold text-center mb-16 text-gray-900 tracking-tight"
    >
      Our Latest Posts
    </h1>
    <ul
      class="grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 gap-12"
    >
      <li
        class="bg-white shadow-lg rounded-xl overflow-hidden transform transition-all hover:scale-105 hover:shadow-2xl duration-300 ease-in-out"
      >
        <div
          class="p-6 flex flex-col justify-between h-full"
        >
          <div
            class="flex-grow"
          >
            <h2
              class="text-2xl font-semibold mb-1 text-gray-800"
            >
              First Post
            </h2>
            <p
              class="text-sm text-gray-500 mb-3"
            >
              October 28, 2024
            </p>
            <p
              class="text-gray-600 mb-4 line-clamp-3"
            >
              This is the first post description.
            </p>
          </div>
          <div
            class="flex justify-end mt-auto"
          >
            <button
              class="bg-gray-900 text-white px-5 py-3 rounded-lg hover:bg-gray-700 transition-colors duration-300 ease-in-out"
            >
              Read More
            </button>
          </div>
        </div>
      </li>
      <li
        class="bg-white shadow-lg rounded-xl overflow-hidden transform transition-all hover:scale-105 hover:shadow-2xl duration-300 ease-in-out"
      >
        <div
          class="p-6 flex flex-col justify-between h-full"
        >
          <div
            class="flex-grow"
          >
            <h2
              class="text-2xl font-semibold mb-1 text-gray-800"
            >
              Second Post
            </h2>
            <p
              class="text-sm text-gray-500 mb-3"
            >
              October 27, 2024
            </p>
            <p
              class="text-gray-600 mb-4 line-clamp-3"
            >
              This is the second post description, which is slightly longer.
            </p>
          </div>
          <div
            class="flex justify-end mt-auto"
          >
            <button
              class="bg-gray-900 text-white px-5 py-3 rounded-lg hover:bg-gray-700 transition-colors duration-300 ease-in-out"
            >
              Read More
            </button>
          </div>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
